vividus-engine/src/main/java/org/vividus/annotation/ReplacementMonitor.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2019-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.annotation;
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Formatter;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang3.Validate;
+import org.jbehave.core.configuration.Keywords;
+import org.jbehave.core.model.StepPattern;
+import org.jbehave.core.steps.NullStepMonitor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.vividus.context.RunTestContext;
+
+public class ReplacementMonitor extends NullStepMonitor
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(ReplacementMonitor.class);
+
+    private final RunTestContext runTestContext;
+
+    private final Map<String, String> stepPatternsCache = new ConcurrentHashMap<>();
+
+    public ReplacementMonitor(RunTestContext runTestContext)
+    {
+        this.runTestContext = runTestContext;
+    }
+
+    @Override
+    public void stepMatchesPattern(String step, boolean matches, StepPattern pattern, Method method,
+            Object stepsInstance)
+    {
+        if (matches)
+        {
+            String key = pattern.type() + " " + pattern.annotated();
+            stepPatternsCache.putIfAbsent(key, pattern.resolved());
+        }
+    }
+
+    @Override
+    public void beforePerforming(String step, boolean dryRun, Method method)
+    {
+        getAnnotation(method).ifPresent(r -> {
+            String rawDeprecatedStep = runTestContext.getRunningStory().getRunningSteps().getFirst();
+            Matcher stepMatcher = getStepMatcher(rawDeprecatedStep);
+            Validate.isTrue(stepMatcher != null, "Couldn't find pattern for deprecated step \"%s\"", rawDeprecatedStep);
```

Comment:
isn't this change dangerous? in case if we do not perform some old step as part of our CI it might block our end users

Reply:
> isn't this change dangerous? in case if we do not perform some old step as part of our CI it might block our end users

In fact, in practice, the exception should never ever throws here.
I had warning on this part of the code and decided to process it
But it should be removed so as not to be misleading
![image](https://github.com/vividus-framework/vividus/assets/58571912/ce14e605-8e90-4200-b88f-7dd4bc157c69)